core/ffmpeg/hlsVariantWriteMonitor_test.go

```diff
@@ -0,0 +1,141 @@
+package ffmpeg
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/sirupsen/logrus/hooks/test"
+)
+
+// time.Sleep calls in this file are present to allow for synchronization
+// to reproduce effects consistently.
+
+func Test_hlsVariantWriteMonitor_playlistLagging(t *testing.T) {
+	logger, hook := test.NewNullLogger()
+	moment, _ := time.Parse(time.RFC3339, "2000-01-01T00:00:00Z")
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	monitor := newHlsVariantWriteMonitor(ctx, logger, time.Second*5)
+
+	monitor.VariantPlaylistWritten("hls/0", moment)
+	monitor.SegmentWritten("hls/0", moment)
+	monitor.SegmentWritten("hls/0", moment.Add(time.Second*4))
+	monitor.SegmentWritten("hls/0", moment.Add(time.Second*8))
+
+	time.Sleep(50 * time.Millisecond)
+
+	if len(hook.Entries) != 1 {
+		t.Fatalf("wrong number of log entries\nexpected: 1\n  actual: %d", len(hook.Entries))
+	}
+
+	if hook.LastEntry().Level != logrus.WarnLevel {
+		t.Errorf("incorrect log error level output: %v", hook.LastEntry().Level)
+	}
+
+	expected := `HLS playlist at "hls/0" not updated with recent segments (8s drift)`
+	if hook.LastEntry().Message != expected {
+		t.Errorf("log message does not match\nexpected: %q\n  actual: %q", expected, hook.LastEntry().Message)
+	}
+}
+
+func Test_hlsVariantWriteMonitor_segmentDelay(t *testing.T) {
```

Comment:
If the time between segments has been delayed long enough, I don't count that as a playlist write miss, since it could be from a stream disconnect and then reconnect.